working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+# Sound non-nullable (by default) types with incremental migration 
+
+Author: leafp@google.com
+
+Status: Draft
+
+This is the proposed specification for [sound non-nullable by default types](http://github.com/dart-lang/language/issues/110).
+Discussion of this proposal should take place in [Issue 110](http://github.com/dart-lang/language/issues/110).
+
+Discussion issues on specific topics related to this proposal are [here](https://github.com/dart-lang/language/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3Annbd+)
+
+The motivations for the feature along with the migration plan and strategy are
+discussed in more detail in
+the
+[roadmap](https://github.com/dart-lang/language/blob/master/working/0110-incremental-sound-nnbd/roadmap.md).
+
+This proposal draws on the proposal that Patrice Chalin wrote
+up [here](https://github.com/dart-archive/dart_enhancement_proposals/issues/30),
+and on the proposal that Bob Nystrom wrote
+up
+[here](https://github.com/dart-lang/language/blob/master/resources/old-non-nullable-types.md).
+
+
+## Syntax
+
+The grammar of types is extended to allow any type to be suffixed with a `?`
+(e.g. `int?`) indicating the nullable version of that type.
+
+A new primitive type `Never`.  This type is denoted by the built-in type
+declaration `Never`, with the same syntactic and scoping treatment as with other
+built-in types such as `Null`.
+
+The grammer of expressions is extended to allow any expression to be suffixed
+with a `!`.
+
+The grammer of instance fields and local variables is extended to allow any
+declaration to include the modifer `lazy`.
+
+The grammar of function types is extended to allow an additional group of named
+parameters prefaced with the word `required` (e.g. `int Function(int, {int?
+y}, required {int z})`.
```

Comment:
Should the above be changed to "allow the `required` modifier before each named parameters"
and `{required int z}` ?

Reply:
Yes.

Reply:
For consistency, with `@required`, I recommend that the new `required` modifier come before all other modifiers for that parameter (e.g. `required final int x`). Sound reasonable?

Reply:
Done.  Left the question of keyword vs built-in identifier as a TODO.